Take the PathBase into account.

jtkech opened the pull request:
Fixes #2891

Jetski5822 commented on src/OrchardCore.Modules/OrchardCore.Tenants/Controllers/AdminController.cs:
```diff
@@ -274,7 +274,7 @@ public async Task<IActionResult> Create(EditTenantViewModel model)
                 shellSettings.Name = model.Name;
                 shellSettings.RequestUrlHost = model.RequestUrlHost;
-                shellSettings.RequestUrlPrefix = model.RequestUrlPrefix?.Trim();
+                shellSettings.RequestUrlPrefix = model.RequestUrlPrefix?.Trim(' ', '/');
```
im wondering - because we do this in 3 places, it might be worth creating a method on the model, or doing this within the getter on the model

jtkech replied:
Okay, i will work on it this night.

Note: i forgot to remove the one in `ModularBackgroundService`, i wanted to keep it only in 2 places (as before with only `.Trim()`), only when we update it from a model (2 different models).

Hmm, finally maybe better to do this when we get it from the shell settings because now the value might come from configuration, a preconfigured value which might have spaces or `/`.

sebastienros commented:
Please don't merge yet and give me a change to understand the issue and this fix. Just to be sure we are doing the correct thing. Thank you.

jtkech commented:
Okay, no problem.

For infos

- Yesterday i tried it under an IIS virtual folder named `orchard`, all seemed to work well.

- Then, @awyl gave me the idea on gitter, i also tried it by using `app.Map()` (see below). In this specific case, all seems to work with the `Default` tenant, i could also create and setup a new tenant. The only thing is that i was not able to login on this new tenant. Let me know if it is worth to fix it for this specific case, if so, because it is related to authentication, i may need the help of @PinpointTownes.

      public void Configure(IApplicationBuilder app, IHostingEnvironment env)
      {
          app.Map("/orchard", builder =>
          {
              if (env.IsDevelopment())
              {
                  builder.UseDeveloperExceptionPage();
              }

              builder.UseStaticFiles();

              builder.UseOrchardCore();
          });
      }

jtkech commented:
- In fact it is the same when creating an IIS virtual folder `/orchard` or using `app.Map("/orchard", ...`. All seems to work, routes, static files paths and so on. The only thing that doesn't work is `Login` on a tenant which has a non null prefix. So it may also affect the `Default` tenant if it is prefixed.

- I could get it working by using in `OC.Users.Startup` the following.

      // In the Startup constructor

      // 'PathBase' doesn't include a tenant prefix when 'Startup' is resolved to call 'ConfigureServices()'.
      _pathBase = httpContextAccessor.HttpContext?.Request.PathBase ?? new PathString(String.Empty);

      services.ConfigureApplicationCookie(options =>
      {
          ...
          options.Cookie.Path = _pathBase + _tenantPrefix;
          ...
      });

I'm working on.

PinpointTownes commented:
@jtkech that's not the first time I hear that and I'm pretty sure @sebastienros mentioned that at some point. I'm a bit surprised we have to manually prepend the cookie path with the path base, as it's something the cookie handler is supposed to do for you at runtime, using a special request-aware cookie builder named `RequestPathBaseCookieBuilder`: https://github.com/aspnet/AspNetCore/blob/master/src/Security/Authentication/Core/src/Internal/RequestPathBaseCookieBuilder.cs#L25

It would be worth checking whether the "original path" stored in the `IAuthenticationFeature` contains the correct path base. If it doesn't, this may indicate that we don't call `app.UseAuthentication()` at the right moment, since `AuthenticationMiddleware` is the component that captures the path base and stores in the feature before invoking the rest of the pipeline: https://github.com/aspnet/AspNetCore/blob/master/src/Security/Authentication/Core/src/AuthenticationMiddleware.cs#L37

jtkech commented:
Okay, thanks for the infos i will check this.

Then i will ask you another question related to antiforgery.

jtkech commented:
@PinpointTownes

Seems that `UseAuthentication()` is called in the right place at the tenant level, and when the middleware is executed **we already have updated `PathBase`** so that `OriginalPathBase` include the tenant prefix and may begin by e.g a virtual folder.

But seems that the cookie builder doesn't take into account `OriginalPathBase` if we have set its `Path` property. So it works when we set it to the tenant prefix which is equal to `PathBase` if no virtual folder. But it doesn't work if under a virtual folder.

**Seems to work if we just don't set the cookie builder path** property so that it uses `OriginalPathBase` which may start with a virtual folder, and already ends with the tenant prefix.

Need more tests.

Meanwhile i have another question.

    private static void AddAntiForgery(OrchardCoreBuilder builder)
    {
         // Commented out for testing so that the below delegate is called
        // builder.ApplicationServices.AddAntiforgery();

        builder.ConfigureServices((services, serviceProvider) =>
        {
            var settings = serviceProvider.GetRequiredService<ShellSettings>();

            var tenantName = settings.Name;
            var tenantPrefix = "/" + settings.RequestUrlPrefix;

            services.AddAntiforgery(options =>
            {
                // This delegate is never called if we call `AddAntiforgery()` at the app level
                options.Cookie.Name = "orchantiforgery_" + tenantName;
                options.Cookie.Path = tenantPrefix;
            });
        });
    }

So, the tenant level delegate is called only if we don't call `AddAntiforgery()` at the app level, then, only in this case, i could see the creation of the related cookies per tenant, otherwise not.

jtkech commented:
@PinpointTownes 

So i changed the way i fixed the authentication part by just not setting the cookie builder path, when you will have time see my above comments and revisew it.

Then, i will dig into the antiforgery part, right now it seems that our `AddAntiForgery()` at the tenant level (whose delegate is never called) is useless, no tenant level antiforgery services are used.

Looking at the aspnetcore code, i think it is because `AddAntiForgery()` uses the `TryAdd` pattern, so the app level one always wins, so maybe we would need to start with a new service collection as we do for `AddDataProtection()`.

PinpointTownes commented on src/OrchardCore.Modules/OrchardCore.Users/Startup.cs:
```diff
@@ -90,8 +82,12 @@ public override void ConfigureServices(IServiceCollection services)
+                // Don't set the cookie builder 'Path' so that it uses the 'IAuthenticationFeature' value
+                // set by the pipeline and comming from the request 'PathBase' which already ends with the
+                // tenant prefix but may also start by a path related e.g to a virtual folder.
+                // options.Cookie.Path = _tenantPrefix;
```
Should that be removed? (just this line, not the precious comment)

jtkech replied:
Okay

PinpointTownes commented on src/OrchardCore/OrchardCore/Modules/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -157,20 +156,23 @@ private static void AddAntiForgery(OrchardCoreBuilder builder)
+                        // set by the pipeline and comming from the request 'PathBase' which already ends with the
+                        // tenant prefix but may also start by a path related e.g to a virtual folder.
+
+                        // options.Cookie.Path = tenantPrefix;
```
Same remark, I don't think it's worth having commented code that uses variables that no longer exist :smile:

jtkech replied:
Okay

PinpointTownes commented on src/OrchardCore/OrchardCore/Modules/Extensions/ServiceCollectionExtensions.cs:
```diff
@@ -157,20 +156,23 @@ private static void AddAntiForgery(OrchardCoreBuilder builder)
-                var pathBase = httpContextAccessor.HttpContext?.Request.PathBase ?? new PathString(String.Empty);
+                // Re-register the antiforgery  services to be tenant-aware.
+                var collection = new ServiceCollection()
+                    .AddAntiforgery(options =>
```
Looks like we had to adopt this pattern in a few places already (data protection, authentication and now antiforgery). Would it be worth reconsidering how tenant services are registered? Would it be sensible to initially inject an empty `IServiceCollection` in the module startups and then, only add the host services to the tenant services collection if no service type of the same type was registered by a module?

Not saying it's a better approach, just something that could be explored/debated.

jtkech replied:
Good idea, i will check if there is no side effects, e.g if there is some places where we really want to add a service only if it is not already registered, even it has been registered at the host level.

PinpointTownes replied:
A much more radical approach would be to consider that tenants have their own services that are completely separated from the host services (i.e tenant services wouldn't inherit from the services defined by the host). That's pretty much what EF Core - that maintains an internal `IServiceProvider` for each `DbContext` - does.

jtkech replied:
So that a tenant becomes a fully isolated "application". Hmm, i'm just thinking about all the services registered e.g by the `WebHostBuilder`.

PinpointTownes replied:
Yeah, if we went that far, we'd likely have to "copy" some common services, like `IHostingEnvironment`, `IConfiguration`, `ILoggerFactory`, `IApplicationLifetime`, `DiagnosticSource`/`DiagnosticListener`. We could also have a `OrchardCoreBuilder.PreserveHostService<TService>()` helper allowing to "transfer" host services.

PinpointTownes replied:
Actually, this reminds me of a recent discussion where I was wondering if it wasn't better if OrchardCore was no longer integrated into ASP.NET Core's Startup class. Instead, it could its own "host", that you would configure using the generic host builder (which will replace `WebHost` in 3.0):

```csharp
public static void Main(string[] args)
{
    CreateHostBuilder(args).Build().Run();
}

public static IHostBuilder CreateHostBuilder(string[] args) =>
    Host.CreateDefaultBuilder(args)
        // Register an OrchardCore ASP.NET Core pipeline:
        .ConfigureOrchardCoreWebHost(builder =>
        {
            // ...
        })

        // You can also configure a regular ASP.NET Core 3.0 app if you want to.
        .ConfigureWebHostDefaults(builder =>
        {
            builder.UseStartup<Startup>();
        });
```

At first, it might seem harsh to hide all the host `IApplicationBuilder` machinery, but in the context of OC, it would probably be a good thing (at least, IMHO).

PinpointTownes commented:
Excellent work, as always.

jtkech commented:
Thanks.
